presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/LogFilter.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.teradata.prestomanager.agent;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayDeque;
+import java.util.ConcurrentModificationException;
+import java.util.Deque;
+import java.util.StringJoiner;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Class for filtering the contents of log files
+ *
+ * Uses a regular expression to match log entries, and treats
+ * non-matching lines as part of the entry on the previous line.
+ *
+ * @see Builder
+ */
+public class LogFilter
+{
+    private final Path file;
+    private final Pattern logPattern;
+    private final String lineSeparator;
+    private final ImmutableMap<String, Predicate<String>> namedGroupFilters;
+    private final ImmutableMap<Integer, Predicate<String>> numberedGroupFilters;
+    private final Integer maxEntries;
+    private final boolean keepFirst;
+
+    private LogFilter(Path file, Pattern logPattern, String lineSeparator,
+            ImmutableMap<String, Predicate<String>> namedGroupFilters,
+            ImmutableMap<Integer, Predicate<String>> numberedGroupFilters,
+            Integer maxEntries, boolean keepFirst)
+            throws FileNotFoundException
+    {
+        this.file = requireNonNull(file);
+        this.logPattern = requireNonNull(logPattern);
+        this.lineSeparator = requireNonNull(lineSeparator);
+        this.namedGroupFilters = requireNonNull(namedGroupFilters);
+        this.numberedGroupFilters = requireNonNull(numberedGroupFilters);
+        this.maxEntries = maxEntries; // null -> no maximum
```

Comment:
If a maximum is provided, I allocate a Deque with that much space. A Deque with MAX_VALUE elements would not work out well.

But I suppose probably not the best idea to pre-allocate the Deque like that anyway, in case someone manually gives a significantly large number. I'll change that.